Foundation/Network/HostName.hsc

```diff
@@ -0,0 +1,166 @@
+-- avoid a race condition and should be safe to use.
+--
+getHostNameInfo :: (Show address_type, Eq address_type, Address address_type)
+                => HostName
+                -> IO (HostNameInfo address_type)
+getHostNameInfo = getHostNameInfo_ Proxy
+
+globalMutex :: MVar ()
+globalMutex = unsafePerformIO (newMVar ())
+
+-- | like `getHostNameInfo` but takes a `Proxy` to help with the type checker.
+getHostNameInfo_ :: (Show address_type, Eq address_type, Address address_type)
+                 => Proxy address_type
+                 -> HostName
+                 -> IO (HostNameInfo address_type)
+getHostNameInfo_ p h@(HostName hn) =
+    withMVar globalMutex $ \_ ->
+    withCString (toList hn) $ \cname -> do
+        ptr <- loop $ c_gethostbyname2 cname (familyCode p)
+
+        cptr <- peek $ castPtr (offname_ptr ptr)
+        on <- getCString cptr
+
+        alptr <- peek $ castPtr (aliases_ptr ptr)
+        alptr' <- peek alptr
+        as <- getAliases alptr'
+
+        ty <- peek $ castPtr (addtype_ptr ptr)
+        len <- peek $ castPtr (length_addr ptr)
+        lptr <- peek $ castPtr (addr_list ptr)
+        lptr' <- peek lptr
+        addrs <- getAddresses p ty len lptr'
+        return $ HostNameInfo (HostName on) as addrs
+  where
+    loop f = do
+        ptr <- f
+        if ptr /= nullPtr
+            then return ptr
+            else do
+                err <- c_get_h_errno
+                case err of
+                    _ | err == (#const HOST_NOT_FOUND) -> throwIO $ HostNotFound h
+                      | err == (#const TRY_AGAIN)      -> loop f
+                      | err == (#const NO_RECOVERY)    -> throwIO FatalError
+                      | err == (#const NO_DATA)        -> throwIO $ NoAssociatedData h
+                      | otherwise                      -> throwIO $ UnknownError err
+    offname_ptr = (#ptr struct hostent, h_name)
+    aliases_ptr = (#ptr struct hostent, h_aliases)
+    addtype_ptr = (#ptr struct hostent, h_addrtype)
+    length_addr = (#ptr struct hostent, h_length)
+    addr_list   = (#ptr struct hostent, h_addr_list)
+
+getCString :: Ptr CChar -> IO String
+getCString ptr = fromList <$> peekCString ptr
+
+getAliases :: Ptr CChar -> IO [HostName]
+getAliases ptr
+    | nullPtr == ptr = return []
+    | otherwise = do
+         x <- HostName <$> getCString ptr
+         (:) x <$> getAliases (ptr `plusPtr` 1)
+
+getAddresses :: (Show address_type, Eq address_type, Address address_type)
+             => Proxy address_type
+             -> Int32
+             -> Int32
+             -> Ptr address_type
+             -> IO [address_type]
+getAddresses p ty len ptr = do
+    x <- peek ptr
+    if x == any
+        then return []
+        else (:) x <$> getAddresses p ty len (ptr `plusPtr` 1)
+
+foreign import ccall unsafe "gethostbyname2"
+    c_gethostbyname2 :: CString -> CInt -> IO (Ptr Word8)
+foreign import ccall unsafe "get_h_errno"
```

Comment:
not really, it is not the same error code (`errno /= h_errno`). I will move the code in the Foundation's Sys binding.